#  🛠️ From SQLite to BigQuery and Looker Studio

## 🔍 Overview:
Welcome to this hands-on data engineering assignment! Here, you’ll learn to work with real-world tools and platforms used by data professionals. 

This assignment is a continuation of a previous analysis where we used SQLite to store COVID-19 outcomes data. Now, we’ll take it a step further by:
1. **Migrating Data**: Moving the SQLite data into Google BigQuery, a robust cloud-based data warehouse.
2. **Setting Up Cloud Services**: Configuring Google Cloud Platform (GCP) with service accounts, authentication, and permissions.
3. **Visualizing Insights**: Using Looker Studio to create an interactive and insightful dashboard.

## 🎯 Learning Goals:
By the end of this assignment, you will:
1. Understand how to migrate data from a local SQLite database to BigQuery.
2. Learn to configure and use GCP, including APIs and service accounts.
3. Create interactive dashboards and visualizations in Looker Studio.
4. Gain confidence in working with cloud-based analytics tools.

---

## 📋 Prerequisites:
Before starting, ensure you have:
1. A Google account to access GCP and Looker Studio.
2. Python installed on your system (version 3.8 or higher).
3. Basic knowledge of:
   - Python scripting
   - SQL queries
   - Cloud platforms (optional but helpful)

---

## 🗂️ Assignment Workflow:
Here’s what we’ll cover step by step:
1. **Set up Environment** 
   - Update python to 3.8 or above.
   - Set up dependencies and install required packages.
2. **Set Up Google Cloud Platform (GCP)**:
   - Create a GCP project.
   - Enable required APIs (BigQuery, Cloud Storage).
   - Set up service accounts and authentication.
3. **Data Migration**:
   - Export SQLite data to  pandas dataframe.
   - Load the data into BigQuery.
4. **Visualization**:
   - Connect BigQuery to Looker Studio.
   - Design interactive dashboards.
5. **Wrap-Up**:
   - Tips,Sharing, troubleshooting.
---

## 🔗 References
- [Google Cloud Platform (GCP)](https://cloud.google.com/)
- [SQLite Documentation](https://www.sqlite.org/docs.html)
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs)
- [Looker Studio Overview](https://lookerstudio.google.com/)
- [Python SQLite Guide](https://docs.python.org/3/library/sqlite3.html)
- [Getting Started with BigQuery](https://cloud.google.com/bigquery/docs/quickstarts)
- [GCP Console](https://console.cloud.google.com/)

---
## 🎉 Let’s Begin!
# Step 1: Setting up Python environment
Let's take the necessary steps to make sure that your Python environment is properly configured and all required dependencies are installed and ready.

In [1]:
# Step 1: Check Python Version and Environment
import sys

print(f"✅ Your Python version: {sys.version}")

# Ensure compatibility
if sys.version_info >= (3, 8):
    print("🎉 Your Python version is compatible!")
else:
    print("❌ Your Python version is lower than 3.8. Please upgrade Python to version 3.8 or higher.")
    print("Follow this guide: https://realpython.com/installing-python/")


✅ Your Python version: 3.14.2 (tags/v3.14.2:df79316, Dec  5 2025, 17:18:21) [MSC v.1944 64 bit (AMD64)]
🎉 Your Python version is compatible!


## Output:
1. If your Python version is **3.8 or higher**: You’re good to go! ✅
2. If your Python version is lower than 3.8:
   - You need to upgrade Python. Follow this guide: [Upgrade Python](https://realpython.com/installing-python/).

---

### Troubleshooting Python Issues:
1. **Using Anaconda**:
   - Check your environment's Python version:
     ```bash
     conda list python
     ```
   - Update Python in your environment:
     ```bash
     conda install python=3.8
     ```
2. **Using Standard Python**:
   - Download the latest version of Python from the [official site](https://www.python.org/downloads/).
   - Install it and ensure it’s added to your system PATH.


In [2]:
import subprocess
import sys

# List of required libraries
required_libraries = ["pandas", "google-cloud-bigquery", "google-cloud-storage", "google-auth", "google-cloud-bigquery"]

# Function to check and install missing libraries
def check_and_install_library(library):
    try:
        __import__(library.split("-")[0])  # Check if the library is installed
        print(f"✅ Library '{library}' is already installed.")
    except ImportError:
        print(f"❌ Library '{library}' is missing.")
        print(f"Installing '{library}'...")
        try:
            subprocess.check_call([sys.executable, "-m", "pip", "install", library])
            print(f"🎉 Successfully installed '{library}'!")
        except Exception as e:
            print(f"🚨 Error installing '{library}': {e}")
            print("👉 Try running the following command in your terminal:")
            print(f"pip install {library}")

# Check and install libraries
for lib in required_libraries:
    check_and_install_library(lib)


✅ Library 'pandas' is already installed.
❌ Library 'google-cloud-bigquery' is missing.
Installing 'google-cloud-bigquery'...
🎉 Successfully installed 'google-cloud-bigquery'!
✅ Library 'google-cloud-storage' is already installed.
✅ Library 'google-auth' is already installed.
✅ Library 'google-cloud-bigquery' is already installed.


## 🖥️ Understanding the Output:
1. ✅ If a library is installed, you’ll see a success message.
2. ❌ If a library is missing, it will be installed automatically.

---

### Troubleshooting Missing Libraries:
1. **Permission Issues**:
   - Add `--user` to the install command:
     ```bash
     pip install google-cloud-bigquery --user
     ```
2. **Environment Issues**:
   - Ensure you’re in the correct environment if using Anaconda:
     ```bash
     conda activate <your-environment-name>
     ```
3. **SSL Certificate Errors**:
   - Upgrade `pip` to fix SSL issues:
     ```bash
     pip install --upgrade pip
     ```
4. **Force Install**:
   - Reinstall the library with this command:
     ```bash
     pip install --force-reinstall google-cloud-bigquery
     ```

Still facing issues? Refer to these official library links for more help:
- [pandas Documentation](https://pandas.pydata.org/)
- [Google Cloud BigQuery Docs](https://cloud.google.com/bigquery/docs)
- [Google Cloud Storage Docs](https://cloud.google.com/storage/docs)


In [3]:
missing_libraries = []
for library in required_libraries:
    try:
        __import__(library.split("-")[0])
        print(f"✅ Library '{library}' is successfully installed.")
    except ImportError:
        missing_libraries.append(library)
        print(f"❌ Library '{library}' is still missing.")

if missing_libraries:
    print("\n🚨 The following libraries are still missing:")
    print(f"pip install {' '.join(missing_libraries)}")
else:
    print("\n🎉 All libraries are installed and verified!")


✅ Library 'pandas' is successfully installed.
✅ Library 'google-cloud-bigquery' is successfully installed.
✅ Library 'google-cloud-storage' is successfully installed.
✅ Library 'google-auth' is successfully installed.
✅ Library 'google-cloud-bigquery' is successfully installed.

🎉 All libraries are installed and verified!


# Step 2: Setting Up Google Cloud Platform (GCP)

## Why Set Up GCP?
To use BigQuery, Google Cloud Platform (GCP) is essential. It allows us to:
1. Organize resources using a **GCP Project**.
2. Enable the **BigQuery API** to interact with our data warehouse.
3. Use a **Service Account** for secure authentication when uploading data.

### 🔗 References:
- [Google Cloud Documentation](https://cloud.google.com/docs)
- [What is BigQuery?](https://cloud.google.com/bigquery/docs/introduction)
- [How to Use Service Accounts](https://cloud.google.com/iam/docs/understanding-service-accounts)
- [BigQuery Pricing](https://cloud.google.com/bigquery/pricing) (to understand usage costs)


## 🌐 Creating a GCP Project

A **GCP Project** is like a folder where you keep all your Google Cloud resources (e.g., APIs, datasets). It helps in organizing and managing access, billing, and usage.

📍 You’ll do all of this from the [BigQuery Console](https://console.cloud.google.com/bigquery).

### How to Create a GCP Project:
1. Go to the [BigQuery Console](https://console.cloud.google.com/bigquery).
2. Log in with your Google account.
3. In the top navigation bar, click **"Select a project"**.
4. Click **"New Project"**.
5. Fill out the form:
   - **Project Name**: Choose something meaningful, e.g., `SQLite-to-BigQuery-Project`.
   - **Organization**: If prompted, leave this as default.
6. Click **"Create"**.

### Output:
1. After creating the project, you’ll see it listed in your project dropdown(Select a project).
2. Note down your **Project ID** (you’ll use it later).
---

### 🔗 References:
- [How to Create a GCP Project](https://cloud.google.com/resource-manager/docs/creating-managing-projects)
- [GCP Project Overview](https://cloud.google.com/resource-manager/docs/overview)


## 📡Enabling BigQuery API

### Why Enable the API?
An **API** (Application Programming Interface) allows tools like Python to interact with GCP services. We need to enable the **BigQuery API** to:
1. Create tables in BigQuery.
2. Upload data from Python scripts.
---
### [Big Query](https://console.cloud.google.com/apis/library/bigquery.googleapis.com) should be enabled by default, if its not follow these steps

### How to Enable BigQuery API:
1. Open the [Google Cloud Console API Library](https://console.cloud.google.com/apis/library).
2. Ensure your project is selected in the top navigation bar.
3. Search for **BigQuery API** in the search bar.
4. Click on **BigQuery API** and then click **Enable**.
Note: It should be enabled by default if its not follow the below steps


### Output:
1. The BigQuery API will appear in your project’s **Enabled APIs and Services** list.
2. You can confirm this by visiting the [API Dashboard](https://console.cloud.google.com/apis/dashboard).

---

### 🔗 References:
- [BigQuery API Documentation](https://cloud.google.com/bigquery/docs/reference/rest)
- [How to Enable APIs in GCP](https://cloud.google.com/apis/docs/enable-disable-apis)

## 🚀 How to Create a Dataset in BigQuery

After enabling BigQuery for your Google Cloud project, follow these steps to create a dataset.



1. Navigate to the **[BigQuery Console](https://console.cloud.google.com/bigquery)**.  
   - This is where you can manage datasets and queries.

2. Select Your Project
   - In the left-hand Explorer panel, find your project (e.g., `your-project-id`).
   - If your project isn’t visible, click the dropdown at the top-left and select your project.

3. Click on the three dots next to your project id(`your-project-id`) and click on **"Create dataset"**

4. Fill in the Dataset Details
   - **Dataset ID**: Enter a name for your dataset (e.g., `nyc_covid_outcomes`). This name will be used to reference your dataset in code and queries.  
   - **Data location**: Choose the location (e.g., `US`) under the Multi-region. This defines where your data will be stored.  

5. Click "Create Dataset"

6. You're Done!
   - Your dataset will now appear under your project in the Explorer panel.  
  You can now upload data or create tables within it.


### 🔗 References:
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs) – Official guide to BigQuery features.
- [Python BigQuery SDK](https://cloud.google.com/python/docs/reference/bigquery/latest) – For Python-based interactions with BigQuery.


   ## 🔑 Setting Up a Service Account

### What is a Service Account?
A **Service Account** is like a digital assistant that allows your Python script to interact with GCP securely. It’s safer than using personal credentials.

---

### How to Create a Service Account:
1. Go to the [IAM & Admin Console](https://console.cloud.google.com/iam-admin/serviceaccounts).
2. Click **"Create Service Account"**.
3. Fill out the details:
   - **Service Account Name**: Use something descriptive like `bigquery-access`.
   - Click **Create and Continue**.
4. Assign the following roles:
   - **BigQuery Admin**: Full access to create and manage BigQuery tables.
   - Search or Scroll down to the BigQuery section and find the BigQuery Admin role.
5. Click **Done**.

---

### Download the Authentication Key:
1. In the Service Account list, click on the newly created Service Account.
2. Navigate to the keys section from the navbar.
2. Click **"Add Key" > "Create New Key"**.
3. Select **JSON** format.
4. Save the key file securely, e.g., `bigquery-key.json`. This is your key file for authentication.

---

### 🔗 References:
- [How to Create a Service Account](https://cloud.google.com/iam/docs/creating-managing-service-accounts)
- [Best Practices for Service Accounts](https://cloud.google.com/iam/docs/best-practices-service-accounts)
- [How to Manage Service Account Keys](https://cloud.google.com/iam/docs/creating-managing-service-account-keys)


In [17]:
#Verify GCP Authentication

from google.oauth2 import service_account
from google.cloud import bigquery

# Replace the key path variable with the path to your downloaded JSON key file
key_path = r"../bigquery-key.json"

try:
    # Authenticate using the Service Account
    credentials = service_account.Credentials.from_service_account_file(key_path)
    client = bigquery.Client(credentials=credentials, project=credentials.project_id)
    print(f"🎉 Successfully authenticated! Project ID: {credentials.project_id}")
except Exception as e:
    print(f"❌ Authentication failed: {e}")
    print("👉 Troubleshooting Tips:")
    print("- Ensure the key file path is correct.")
    print("- Verify the Service Account has the BigQuery Admin role.")
    print("- Confirm the BigQuery API is enabled for your project.")

🎉 Successfully authenticated! Project ID: daen-328-assignments


## 🖥️ Understanding the Output:
1. **Success**:
   - If everything is set up correctly, you’ll see a confirmation with your Project ID.
2. **Failure**:
   - Authentication issues may occur if:
     - The JSON key file path is incorrect.
     - The Service Account lacks the correct roles.
     - The BigQuery API is not enabled.

---

## 🔧 Troubleshooting:
1. **Key File Issues**:
   - Ensure the JSON key file is in the correct location and hasn’t been renamed.
2. **Role Issues**:
   - Add the BigQuery Admin role to your Service Account using the [IAM Console](https://console.cloud.google.com/iam-admin/iam).
3. **API Issues**:
   - Revisit the [API Library](https://console.cloud.google.com/apis/library) and confirm the BigQuery API is enabled.

---

### 🔗 References:
- [Google Cloud Authentication](https://cloud.google.com/docs/authentication/production)
- [BigQuery Client Library Documentation](https://cloud.google.com/python/docs/reference/bigquery/latest)


# Step 3: Uploading Data from SQLite to BigQuery

## Why Upload Data to BigQuery?
BigQuery is a powerful data warehouse that allows us to:
1. Store large datasets in the cloud.
2. Query data using SQL with incredible speed and scalability.
3. Integrate with tools like Looker Studio for visualization.

In this step, we’ll:
1. Extract data from an SQLite database.
2. Transform it into a format BigQuery understands(Pandas DataFrame).
3. Upload it directly to a BigQuery table.

---

## 🔗  References
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs)
- [Google BigQuery Python Client Library](https://cloud.google.com/python/docs/reference/bigquery/latest)
- [SQLite Documentation](https://www.sqlite.org/docs.html)

In [18]:
#Import Required Libraries and Set Up Configuration
import pandas as pd
from sqlite3 import connect
from google.cloud import bigquery
from google.oauth2 import service_account

# Set up the path to your SQLite database and BigQuery credentials
SQLITE_DB_FILE = r"../A1/nyc_covid_outcomes.db"  # Replace with your SQLite database file from the first assignment
BIGQUERY_KEY_FILE = r"../bigquery-key.json"  # Replace with your Service Account JSON key file that you just downloaded
PROJECT_ID = "daen-328-assignments"  # Replace with your GCP Project ID
DATASET_ID = "nyc_covid_outcomes"  # Replace with your BigQuery dataset ID(this is the name of the dataset in the previous step)
# PROJECT_ID: 
# - Found in the GCP Console (https://console.cloud.google.com)
# - Navigate to the top-left corner, select the project (id must be displayed next to the project name or on the homescreen)
# DATASET_ID:
# - Found under "BigQuery" in the GCP Console.
# - Navigate to the BigQuery section (https://console.cloud.google.com/bigquery).
# - Select your project from the left panel, then select your dataset to view the dataset info 

###  Loading Data from SQLite:
1. Connect to the SQLite database.
2. Use a SQL query to extract all records from the `covid_outcomes` table.
3. Load the extracted data into a Pandas DataFrame for further processing.

In [19]:
# Load Data from SQLite into a Pandas DataFrame
def load_data_from_sqlite():
    """
    Load data from the 'covid_outcomes' table into a Pandas DataFrame.
    
    Returns:
        pd.DataFrame: Data loaded from the SQLite table.
    """
    try:
        # Connect to the SQLite database
        conn = connect(SQLITE_DB_FILE)
        query = "SELECT * FROM covid_outcomes"
        print("🔄 Loading data from the 'covid_outcomes' table...")
        data = pd.read_sql_query(query, conn)
        conn.close()
        print(f"✅ Successfully loaded {len(data)} records from 'covid_outcomes'.")
        return data
    except Exception as e:
        print(f"❌ Error loading data from 'covid_outcomes': {e}")
        return pd.DataFrame()


### Uploading Data to BigQuery
1. Authenticate with BigQuery using the Service Account key.
2. Define a schema for the BigQuery table (if it doesn’t already exist).
3. Upload the data from the DataFrame into the BigQuery table.


In [20]:
#Upload DataFrame to BigQuery
def upload_to_bigquery(dataframe, table_name):
    """
    Upload a Pandas DataFrame to a BigQuery table.
    
    Args:
        dataframe (pd.DataFrame): The DataFrame to upload.
        table_name (str): The name of the BigQuery table (e.g., "your_dataset_id.table_name").
    """
    try:
        # Authenticate and initialize the BigQuery client
        credentials = service_account.Credentials.from_service_account_file(BIGQUERY_KEY_FILE)
        client = bigquery.Client(credentials=credentials, project=PROJECT_ID)

        # Define table ID
        table_id = f"{PROJECT_ID}.{DATASET_ID}.{table_name}"
        
        # Configure the load job
        job_config = bigquery.LoadJobConfig(
            write_disposition=bigquery.WriteDisposition.WRITE_APPEND,  # Append to existing table
            create_disposition=bigquery.CreateDisposition.CREATE_IF_NEEDED  # Create table if not exists
        )

        # Start the upload job
        print(f"🔄 Uploading {len(dataframe)} records to BigQuery table: {table_id}")
        job = client.load_table_from_dataframe(dataframe, table_id, job_config=job_config)
        job.result()  # Wait for the job to complete
        print(f"✅ Successfully uploaded {len(dataframe)} records to {table_id}")
    except Exception as e:
        print(f"❌ Error uploading to BigQuery: {e}")


In [21]:
# Full Workflow
def push_to_bigquery():
    """
    Full workflow to load data from SQLite and upload to BigQuery.
    """
    print("🔄 Starting the workflow to upload data to BigQuery...")
    # Load data from SQLite
    dataframe = load_data_from_sqlite()
    
    if not dataframe.empty:
        # Specify the table name
        table_name = "covid_outcomes"  # Replace with your desired BigQuery table name
        
        # Upload data to BigQuery
        upload_to_bigquery(dataframe, table_name)
    else:
        print("⚠️ No data found in the 'covid_outcomes' table. Workflow skipped.")

# Execute the workflow
push_to_bigquery()

🔄 Starting the workflow to upload data to BigQuery...
🔄 Loading data from the 'covid_outcomes' table...
✅ Successfully loaded 176148 records from 'covid_outcomes'.
🔄 Uploading 176148 records to BigQuery table: daen-328-assignments.nyc_covid_outcomes.covid_outcomes


c:\Users\annaw\OneDrive\Documents\DAEN328\daen328_env\Lib\site-packages\google\cloud\bigquery\_pandas_helpers.py:486: FutureWarning: Loading pandas DataFrame into BigQuery will require pandas-gbq package version 0.26.1 or greater in the future. Tried to import pandas-gbq and got: No module named 'pandas_gbq'
  warnings.warn(


✅ Successfully uploaded 176148 records to daen-328-assignments.nyc_covid_outcomes.covid_outcomes


## 🖥️ Understanding the Output:
1. **Data Loaded Successfully**:
   - If the SQLite table contains data, you’ll see a message confirming the number of records loaded.
2. **Upload Status**:
   - A message confirms successful upload to BigQuery.
   - If any errors occur, detailed messages will help debug the issue.

---

## 🕵️ Exploring the Uploaded Data in BigQuery:
1. **Verify the Table in BigQuery**:
   - Open the **[BigQuery Console](https://console.cloud.google.com/bigquery)**.
   - Navigate to your project in the left-hand Explorer panel.
   - Expand the dataset where your table is uploaded (e.g., `nyc_covid_outcomes`).
   - Click on the table name (e.g., `covid_outcomes`) to open it.

2. **Preview the Data**:
   - In the table view, click the **"Preview"** tab to view the first few rows of your data. This is a quick way to check if the upload was successful.
   - Check the **"Schema"** tab to verify the column names and data types.
3. **Run Queries**:
   - Click on the **"+"** tab and use SQL to explore your data. Example query:
     ```sql
     SELECT * 
     FROM `your-project-id.nyc_covid_outcomes.covid_outcomes`
     LIMIT 10;
     ```
   - Modify the query to filter or summarize the data as needed.



---


## 🔧 Troubleshooting:
1. **SQLite Issues**:
   - Ensure the database file exists in the specified path.
   - Verify the `covid_outcomes` table exists using a database viewer like DB Browser.
2. **BigQuery Issues**:
   - Ensure the Service Account JSON key is correctly configured.
   - Confirm the dataset exists in BigQuery.
   - Verify the Service Account has the necessary roles (BigQuery Admin).

---

### 🔗 References:
- [BigQuery Python Client](https://cloud.google.com/python/docs/reference/bigquery/latest)
- [Pandas Documentation](https://pandas.pydata.org/docs/)
- [SQLite Documentation](https://www.sqlite.org/docs.html)


# Step 4: Visualizing Data in Looker Studio

## Why Use Looker Studio?
[Looker Studio](https://lookerstudio.google.com/) allows us to:
1. Create interactive dashboards to visualize BigQuery data.
2. Share insights in a user-friendly format.
3. Enable real-time exploration of datasets.

In this step, we will:
1. Connect Looker Studio to BigQuery.
2. Build an interactive dashboard for the `covid_outcomes` table.

---

### 🔗 References:
- [Looker Studio Overview](https://lookerstudio.google.com/)
- [Connecting Looker Studio to BigQuery](https://support.google.com/looker-studio/answer/6370296?hl=en)
- [BigQuery Visualization Best Practices](https://cloud.google.com/bigquery/docs/visualize-looker)


## 📈 Connecting Looker Studio to BigQuery

Connecting Looker Studio to BigQuery allows us to directly query our dataset and create visualizations without exporting the data manually.

### How to Connect:
1. Open [Looker Studio](https://lookerstudio.google.com/).
2. Click **"Blank Report"** or **"Create"**.
3. In the **Add Data to Report** dialog:
   - Select **BigQuery**.
   - Click **Authorize** if prompted to grant access.
4. Choose your project:
   - Select your GCP project (e.g., `SQLite-to-BigQuery-Project`).
   - Navigate to the dataset (e.g., `nyc_covid_outcomes`).
   - Select the table (e.g., `covid_outcomes`).
5. Click **Add** and confirm by clicking **Add to Report**.

---

### Output:
Your BigQuery table will now appear as a data source in Looker Studio. You can start building visualizations using this data source.


## 📊 Creating Visualizations

### Why Create Visualizations?
Visualizations help us understand data better by showing patterns, trends, and comparisons. In this step, we’ll create easy-to-build charts to explore the `covid_outcomes` data in Looker Studio.

---

### Suggested Charts for `covid_outcomes` Table

#### 1. **Confirmed Cases Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of confirmed cases changed over time.
- **Steps**:
  1. Add a **Line Chart** from the toolbar.
  2. Set the **Dimension** to `specimen_date`.
  3. Set the **Metric** to `number_confirmed`.
  4. Rename the chart title to "Confirmed Cases Over Time" under the style tab.

---

#### 2. **Deaths Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of deaths changed over time.
- **Steps**:
  1. Duplicate the "Daily Confirmed Cases" chart (right-click > Duplicate).
  2. Replace the **Metric** with `number_deaths`.
  3. Rename the chart title to "Deaths Over Time."

---
#### 3. **Hospitalizations Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of Hospitalizations changed over time.
- **Steps**:
  1. Duplicate the "Daily Confirmed Cases" chart (right-click > Duplicate).
  2. Replace the **Metric** with `number_hospitalized`.
  3. Rename the chart title to "Hospitalizations Over Time."

---


#### 4. **Key Metrics Summary**
- **Chart Type**: Scorecards
- **Purpose**: Highlight total numbers for key metrics (e.g., total confirmed cases, total deaths).
- **Steps**:
  1. Add **4 Scorecards** from the toolbar.
  2. For each scorecard, set the **Metric** to one of the following:
     - `number_tested` 
     - `number_confirmed` 
     - `number_hospitalized` 
     - `number_deaths` 
  3. Rename the scorecards to "Total Tested," "Total Confirmed," etc.

---
### 5. Testing vs Confirmed Cases
- **Chart Type**: Combo Chart (Line and Bar)
- **Purpose**: Compare the number of tests conducted with confirmed cases over time.

#### Steps:
1. Add a **Combo Chart**.
2. Set the **Dimension** to `specimen_date`.
3. Add the following **Metrics**:
   - `number_tested` (Bar)
   - `number_confirmed` (Line)
4. Rename the chart title to **"Testing vs Confirmed Cases"**.

---
### 6. **Custom Calculations**
- **Purpose**: Create additional metrics to derive more insights from the existing data using calculated fields.
- **Usage**: 
  - In the **Data tab** (on the right-hand side), you can view the current schema of the data being used.
  - To create a custom value, click **Add a Field** → **Add a Calculated Field**.

#### **Examples of Custom Calculations**:
1. **Death Rate**:
   - Formula: `number_deaths / number_confirmed`
2. **Hospitalization Rate**:
   - Formula: `number_hospitalized / number_confirmed`
3. **Positive Test Rate**:
   - Formula: `number_confirmed / number_tested`

#### **Visualization**:
- After creating calculated fields, use them in visualizations like **Line Charts** or **Bar Charts**:
  1. Add a **Line Chart** or **Bar Chart**.
  2. Set the **Dimension** to `specimen_date`.
  3. Use the calculated values (e.g., Death Rate, Hospitalization Rate) as **Metrics**.
  4. Customize the chart title for clarity (e.g., "Death Rate Over Time").

### Output:
By the end of this step, your dashboard should have charts that provide a detailed overview of the COVID-19 data.


## 📤 Finalizing and Sharing

### How to Finalize:
1. Review all charts for accuracy and aesthetics.
2. Customize the layout and design:
   - Add a title and description for your dashboard.
   - Use consistent colors for metrics.
3. Create separate pages for sections of data.
4. Click **View** in the top-right corner to review the report's functionality and visual design.

---

### How to Share:
1. Click **Share** in the top-right corner.
2. Choose one of the following options:
   - **Invite People**: Add specific viewers by email.
   - **Get Link**: Create a shareable link (adjust permissions if needed).

---

## 🎉 Your dashboard is now ready to share and explore!


## 🔧 Troubleshooting Looker Studio

### Common Issues:
1. **BigQuery Data Not Loading**:
   - Ensure your Service Account has access to the dataset.
   - Verify the dataset and table names are correct.
2. **Chart Not Displaying Data**:
   - Check that dimensions and metrics are correctly assigned.
   - Ensure the data type matches the chart type (e.g., numeric for metrics).

---

### 🔗 References:
- [Looker Studio Support](https://support.google.com/looker-studio/)
- [Fixing Common Looker Studio Issues](https://support.google.com/looker-studio/answer/7020039)


# 📝 Step 5: Final Deliverables

You have migrated your data and built your dashboard. This last step is where you show us what it means. **This section is worth 10 points.**

---

## 🔗 Deliverable 3: Your Looker Studio Report Link

Paste your report's share link below. Before you do, two checks:

1. **Sharing**: In Looker Studio, click **Share** → set link access to **"Anyone with the link"** → **Viewer**.
2. **Data credentials**: Open your BigQuery data source → **Data credentials** → set to **Owner's credentials**. If this is left on *Viewer's credentials*, Looker Studio will try to query BigQuery using **our** account when we open your link — and we have no access to your project. Your report will render as an error and score as a report you did not build.

Verify by opening the link in a private/incognito window while signed out of your Google account.

**My Looker Studio report link:** https://datastudio.google.com/s/r8UvezX7urQ

---

## ✍️ Deliverable 4: Written Interpretation

### 👉 The questions are in **`Assignment02_Addendum.pdf`**, attached with this assignment.

That file contains **20 questions** in four sections:

| Section | Topic | Questions |
|---|---|---|
| 1 | GCP & BigQuery Basics | 1–5 |
| 2 | Setting Up BigQuery & Data Upload | 6–11 |
| 3 | Querying Data in BigQuery | 12–16 |
| 4 | Exploring, Validating, and Understanding Data | 17–20 |

### How to submit your answers

1. Open `Assignment02_Addendum.pdf`.
2. Add a **new markdown cell below this one**.
3. Answer **all 20 questions**, numbered to match the addendum so we can follow along.
4. For questions asking for SQL (12–14), format your query in a code block so it stays readable:

   ````
   ```sql
   SELECT COUNT(*) FROM `your-project-id.nyc_covid_outcomes.covid_outcomes`;
   ```
   ````

### What earns full marks

Answer from **your own project, your own numbers, and your own dashboard**. Many of these questions ask what *you* did and what *you* saw — answers that could have been written without doing the assignment will score in the "Generic description not tied to the student's own numbers" band of the rubric.

Two questions carry extra weight, because the rubric requires a data quality observation carried forward from Assignment 1:

- **Q18** — patterns and trends: describe what your charts actually show, with approximate dates, not what you expected to see.
- **Q19** — validation: this is where that Assignment 1 data quality issue belongs. Say whether it still appears in BigQuery and how it affects the way your dashboard should be read.

---

## ✅ Before you submit

- [ ] Row count in BigQuery reconciles with Assignment 1 (re-running the upload cell is the usual cause of a mismatch)
- [ ] All six visuals present, every time axis using `specimen_date` sorted ascending
- [ ] All three calculated fields defined and used in at least one chart
- [ ] All 20 addendum questions answered and numbered
- [ ] Share link opens correctly in a private browser window
- [ ] Notebook run top-to-bottom with outputs visible
- [ ] **No key file and no `.db` file in your GitHub repo** — run `git status` and confirm


---

### GCP & BigQuery Basics

1. IaaS (infrastructure as a service) is the most simplistic form of cloud computing services as it involves renting out the hardware itself, such as Google Compute Engine which allows users to generate and run virtual machines. PaaS (platform as a service) provides the hardware as well as some managed services. For example, Google Cloud SQL is a PaaS that provides relational databases with some built-in automation. SaaS (software as a service) builds even further off the capabilities of a PaaS by providing a complete software platform that provides fully implemented services. An example within GCP is Google Workspace which contains various services such as Gmail, Google Docs, Google Drive, etc. which require zero technical knowledge to use.

2. A cloud-based data warehouse is managed by an external provider that companies or users pay to gain access to a certain amount of storage. This is different from an on-premises database as traditional databases are owned and managed by the company or user themselves and typically only serve that one entity. Cloud-based data warehouses are becoming increasingly popular as companies no longer have to be responsible for managing and maintaining the storage hardware and can instead outsource that labor to the company that owns the cloud-based warehouse.

3. In order to be able to use the BigQuery API, it has to be enabled on the account attempting to use the API. Since the API is how tables are created and data is uploaded when using Python, being able to send API requests to BigQuery is necessary.

4. A service account is a form of GCP account that is intended to be used to access GCP services outside of the GUI, like BigQuery API. On the other hand, a user account is simply how a specific user/login is associated with various projects and can contain multiple service accounts. A service account was therefore necessary for this assignment since I needed to be able to use BigQuery API, and a service account sets the permissions as well as allow for the creation of authentication keys for using the BigQuery API.

5. I used the google-auth library in the python script to create credentials out of the authentication key JSON file that could then be passed as a parameter to the BigQuery API. This ensures that the python script was connecting to the proper BigQuery account and project.

### Setting up BigQuery & Data Upload

6. I created a BigQuery project using the BigQuery Console GUI by pressing Select a Project -> New Project and then providing a name for the new GCP project. Once the project was made, I could create a dataset within the project by navigating to the project in the Explorer area of the GUI, clicking the three dots next to my newly created project, and creating a new dataset with the ID and region of my choice. I then created the table using python through the BigQuery API by establishing a connection to the proper BigQuery project using the authentication key JSON file and the project ID. Next, I created a job with the instructions to create a table if a table with the provided name does not exist, and then passed that job as a parameter within a BigQuery API request. This was a load_table_from_dataframe request with a table ID of daen-328-assignments.nyc_covid_outcomes.covid_outcomes. Since my dataset did not have any tables prior to running this API request, it created a new table with the provided ID and wrote the data from the dataframe into it.

7. The database I uploaded consisted of one table, covid_outcomes, with six columns and 176,148 rows. Of the six columns, four of them were of type integer (number_tested, number_confirmed, number_hospitalized, and number_deaths) while the other two were strings (extract_date and specimen_date). 

8. I used google-cloud-bigquery and google-auth libraries to connect to BigQuery. google-cloud-bigquery assists in establishing a connection to the BigQuery API and handling the requests sent to the API (queries, database creation, etc). google-auth uses credentials (like the BigQuery JSON key) to ensure the code connects to the correct BigQuery service account and project.

9. I had to connect to the SQLite database, use a SQL query to retrieve the entire database, and place the results of the query into a pandas DataFrame. I then disconnected from SQLite and connected to a specific dataset in BigQuery using the BigQuery Service Account key, project ID, and dataset ID. Once I was connected to the BigQuery API, I created a table (if it did not already exist) so I would have a place to upload my DataFrame to. Lastly I uploaded the dataframe to the new BigQuery table and disconnected from the BigQuery API.

10. The only challenge I faced when uploading the data came from establishing the connection using the authentication key. It took a couple of tries to write a relative path that successfully navigated to the intended file so that I could establish a connection to the BigQuery API. Ultimately, I had to print out the current directory so I could accurately determine where the JSON file was in relation to the script's directory.

11. For a much larger dataset, I would likely skip loading the data into a SQLite database since that takes up local storage. Instead, I would just load the data from the API into a dataframe and pass that dataframe directly to BigQuery. This would also save time by avoiding loading data into SQLite, retrieving the data from SQLite, converting it to a dataframe, and then finally sending it to BigQuery. This is especially relevant in large datasets as writing and retrieving millions of rows can take a long time, so it is ideal to do it as little as necessary. I would also add an extra cleaning step to remove rows with a date in the future since there were specimen dates that survived the cleaning/validation in Assignment 1 that made later analysis more difficult.

### Querying Data in BigQuery

12.
````
```sql
SELECT COUNT(*) FROM `daen-328-assignments.nyc_covid_outcomes.covid_outcomes`;
```
````

13.

````
```sql
SELECT * FROM `daen-328-assignments.nyc_covid_outcomes.covid_outcomes` LIMIT 10;
```
````

14. Note: The choice of 2020-08-01 and 2021-08-01 are just placeholders for the start (inclusive) and end (exclusive) of any date range. 

````
```sql
SELECT * FROM `daen-328-assignments.nyc_covid_outcomes.covid_outcomes`
WHERE extract_date >= '2020-08-01' AND extract_date < '2021-08-01';
```
````

15. The purpose of LIMIT in a query is to set a maximum number of rows to return, and it is useful when a specific query would return thousands of rows but you only need to see a subset of them. For example, using LIMIT in a query that is just meant to check that the table structure is correct is beneficial because only a small number of rows are returned as opposed to the entire table.

16. An optimal SQL query retrieves or edits only as much data as is necessary/intended. For example, to validate the structure of the data after loading it into a new database table, limiting the output to 5 or 10 rows is much more optimal than returning every row since you are unlikely to look past the tenth row to see that everything loaded in properly. Queries that return or manipulate more data are likely to cost more and take longer to run on BigQuery because the system has to do more work.

### Exploring, Validating, and Understanding Data

17. I used Looker to visualize various metrics from the BigQuery data so that I could easily examine trends that would be hard to notice by just looking at thousands of rows of raw numbers. It also allowed for comparison of different columns, either through finding total counts (like number of confirmed cases or hospitalized cases) or by plotting more than one metric on the same graph.

18. Overall, the data seemed pretty reasonable based on my analysis. The data appeared to match my assumptions of overall trends with higher numbers of tests, confirmed cases, hospitalizations, and deaths near the start of COVID when hospitals were overwhelmed and doctors were still determining the best course of treatment (March 2020 to May 2020). However, once quarantine and lockdown were enforced, there is a significant drop in cases, and severe cases also reduced around this time as doctors learned more about the virus. The numbers stay relatively low for the remainder of the time frame covered by the data with a small spike for all metrics in December 2020/January 2021, likely due to gatherings for the holidays.

19. Since Assignment 1 already did some validation, I did not do any additional validation on the data while in table form. Instead, I started by examining the plots on Looker Studio to make sure they appeared reasonable and did not have obvious issues (e.g. number of confirmed cases exceeding number of tests). Since the erroneous specimen_date entries were still present in the BigQuery dataset, I filtered every plot to only include specimen dates between January 1, 2020 and October 1, 2021 to focus the plots on valid data. I could not find any obvious issues, so I moved on to comparing the plots with some other sources. 

    I looked at an interactive graph of cases for the same time period created by Our World in Data using data from the World Health Organization (https://ourworldindata.org/covid-cases). This showed fairly similar results with the exception of showing drastically less confirmed cases, hospitalizations, and deaths at the beginning of COVID (March 2020 - May 2020) than the NYC dataset showed. Our World in Data shows the highest number of confirmed cases, hospitalizations, and deaths occuring around January 2021 whereas this dataset shows only a small spike around that time. This dataset has confirmed case numbers from March to May of 2020 that are 3 times larger as well as 30 times the number of hospitalizations/deaths in comparison to January 2021. 

    Due to this discrepancy, I looked at another source provided by the CDC that plotted COVID hospitalization rates over time (https://www.cdc.gov/covid/php/covid-net/index.html). The CDC's data aligned with the data provided by WHO by also showing the magnitude of hospitalizations in January 2021 being larger than in March to May of 2020. This leads me to the conclusion that either the NYC data may be invalid (although the method of data collection seems reliable), or that NYC had better reporting systems in place for tracking cases and their severity early in COVID, leading to higher values. Another factor is that the plots I created do not take into account the limitations of this dataset as they do not exclusively show the latest extract_date data for each specimen_date and instead sum all of the values for a given specimen_date. Since the early months likely had more frequent updates with new patients being hospitalized or dying, those specimen dates may appear more frequently in the dataset and therefore inflate the plots. This seems like the most plausible explanation, though I would also expect that NYC may have had better reporting in comparison to the remainder of the country.

20. I could have skipped uploading the data to SQLite and instead loaded it directly into a dataframe to then send to BigQuery. Although loading directly into BigQuery would reduce the number of steps, it could be easier for mistakes to go unnoticed. By temporarily storing the data in SQLite, I could run basic queries and perform cleaning/validation that would have been slightly harder to do on a dataframe or in BigQuery. However, using SQLite does produce another copy of the dataset and require transferring the data multiple times, which is more time-consuming and also introduces potential for mistakes.

